lib/AST/LayoutConstraint.cpp

```diff
@@ -114,4 +136,61 @@ void *LayoutConstraintInfo::operator new(size_t bytes, const ASTContext &ctx,
 
 SourceRange LayoutConstraintLoc::getSourceRange() const { return getLoc(); }
 
+LayoutConstraint LayoutConstraint::merge(LayoutConstraint Other) {
+  if (!(*this)->isKnownLayout())
+    return Other;
+
+  if ((*this)->isAddressOnlyTrivial() && Other->isKnownSizeTrivial()) {
+    // Prefer a more concrete _Trivial constraint.
+    return Other;
+  }
+
+  if (Other->isAddressOnlyTrivial() && (*this)->isKnownSizeTrivial()) {
+    // Prefer a more concrete _Trivial constraint.
+    return *this;
+  }
+
+
+  if ((*this)->isRefCountedObject() && Other->isNativeRefCounted()) {
+    // Prefer a more concrete _NativeRefCounted constraint.
+    return Other;
+  }
+
+  if (Other->isRefCountedObject() && (*this)->isNativeRefCounted()) {
+    // Prefer a more concrete _NativeRefCounted constraint.
+    return *this;
+  }
+
+  if ((*this)->isClass() && Other->isNativeClass()) {
+    // Prefer a more concrete _NativeClass conttraint.
+  }
+
+  if (Other->isClass() && (*this)->isNativeClass()) {
+    // Prefer a more concrete _NativeClass conttraint.
+  }
+
+  return LayoutConstraint::getUnknownLayout();
```

Comment:
Might it be better to have a 2-D table to describe the results? It's pretty hard to make sure that this long sequence of `if` statements does the right thing.